pkg/cmd/cli/cmd/rollback.go

```diff
@@ -38,151 +43,267 @@ a human-readable representation of the updated deployment configuration instead
+	// Extract basic flags.
+	if len(args) == 1 {
+		o.TargetName = args[0]
+	}
+	namespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	o.Namespace = namespace
+
+	// Set up client based support.
+	mapper, typer := f.Object()
+	o.getBuilder = func() *resource.Builder {
+		return resource.NewBuilder(mapper, typer, f.ClientMapperForCommand())
+	}
 
-// Describe describes a DeploymentConfig.
-func (r *helper) Describe(config *deployapi.DeploymentConfig, out io.Writer) error {
-	description, err := r.describe(config)
+	oClient, kClient, err := f.Clients()
 	if err != nil {
 		return err
 	}
-	out.Write([]byte(description))
+	o.oc = oClient
+	o.kc = kClient
+
+	o.out = out
+	return nil
+}
+
+func (o *RollbackOptions) Validate() error {
+	if len(o.TargetName) == 0 {
+		return fmt.Errorf("a deployment or deploymentconfig name is required")
+	}
+	if o.DesiredVersion < 0 {
+		return fmt.Errorf("the to version must be >= 0")
+	}
 	return nil
 }
 
-// Print prints a deployment config in the specified format with the given
-// template.
-func (r *helper) Print(config *deployapi.DeploymentConfig, format, template string, out io.Writer) error {
-	printer, _, err := kubectl.GetPrinter(format, template)
+func (o *RollbackOptions) Run() error {
+	// Get the resource referenced in the command args.
+	obj, err := o.findResource(o.TargetName)
+	if err != nil {
+		return err
+	}
+
+	// Interpret the resource to resolve a target for rollback.
+	var target *kapi.ReplicationController
+	switch r := obj.(type) {
+	case *kapi.ReplicationController:
+		// A specific deployment was used.
+		target = r
+	case *deployapi.DeploymentConfig:
+		// A deploymentconfig was used. Find the target deployment by the
+		// specified version, or by a lookup of the last completed deployment if
+		// no version was supplied.
+		deployment, err := o.findTargetDeployment(r, o.DesiredVersion)
+		if err != nil {
+			return err
+		}
+		target = deployment
+	}
+	if target == nil {
+		return fmt.Errorf("%s is not a valid deployment or deploymentconfig", o.TargetName)
+	}
+
+	// Set up the rollback and generate a new rolled back config.
+	rollback := &deployapi.DeploymentConfigRollback{
+		Spec: deployapi.DeploymentConfigRollbackSpec{
+			From: kapi.ObjectReference{
+				Name: target.Name,
+			},
+			IncludeTemplate:        true,
+			IncludeTriggers:        o.IncludeTriggers,
+			IncludeStrategy:        o.IncludeStrategy,
+			IncludeReplicationMeta: o.IncludeScalingSettings,
+		},
+	}
+	newConfig, err := o.oc.DeploymentConfigs(o.Namespace).Rollback(rollback)
+	if err != nil {
+		return err
+	}
+
+	// If this is a dry run, print and exit.
+	if o.DryRun {
+		describer := describe.NewDeploymentConfigDescriberForConfig(o.oc, o.kc, newConfig)
+		description, err := describer.Describe(newConfig.Namespace, newConfig.Name)
+		if err != nil {
+			return err
+		}
+		o.out.Write([]byte(description))
```

Comment:
Niggly, but will this leave me with my bash prompt showing up on the same line as the end of the output?  I dislike that.